modules/AlphaWalletFoundation/AlphaWalletFoundation/Tokens/Helpers/NonFungibleErc1155JsonBalanceFetcher.swift

```diff
@@ -20,27 +20,39 @@ class NonFungibleErc1155JsonBalanceFetcher {
     private let erc1155TokenIdsFetcher: Erc1155TokenIdsFetcher
     private let erc1155BalanceFetcher: Erc1155BalanceFetcher
 
-    private let session: WalletSession
+    private let server: RPCServer
     private let tokensService: TokenProvidable
     private let importToken: ImportToken
 
-    init(tokensService: TokenProvidable, session: WalletSession, erc1155TokenIdsFetcher: Erc1155TokenIdsFetcher, jsonFromTokenUri: JsonFromTokenUri, erc1155BalanceFetcher: Erc1155BalanceFetcher, importToken: ImportToken) {
-        self.session = session
+    init(tokensService: TokenProvidable,
+         server: RPCServer,
+         erc1155TokenIdsFetcher: Erc1155TokenIdsFetcher,
+         jsonFromTokenUri: JsonFromTokenUri,
+         erc1155BalanceFetcher: Erc1155BalanceFetcher,
+         importToken: ImportToken) {
+
+        self.server = server
         self.erc1155TokenIdsFetcher = erc1155TokenIdsFetcher
         self.tokensService = tokensService
         self.jsonFromTokenUri = jsonFromTokenUri
         self.erc1155BalanceFetcher = erc1155BalanceFetcher
         self.importToken = importToken
     }
 
+    deinit {
+        print("xxx.\(self).deinit for server: \(server)")
+    }
+
     func fetchErc1155NonFungibleJsons(enjinTokens: EnjinTokenIdsToSemiFungibles) -> AnyPublisher<[AlphaWallet.Address: [NonFungibleBalanceAndItsSource<NonFungibleFromTokenUri>]], SessionTaskError> {
         return erc1155TokenIdsFetcher
             .detectContractsAndTokenIds()
             .mapError { SessionTaskError(error: $0) }
-            .flatMap { contractsAndTokenIds -> AnyPublisher<Erc1155TokenIds.ContractsAndTokenIds, SessionTaskError> in
-                self.addUnknownErc1155ContractsToDatabase(contractsAndTokenIds: contractsAndTokenIds.tokens)
-            }.flatMap { contractsAndTokenIds -> AnyPublisher<(contractsAndTokenIds: Erc1155TokenIds.ContractsAndTokenIds, tokenIdMetaDatas: [TokenIdMetaData]), SessionTaskError> in
-                self._fetchErc1155NonFungibleJsons(contractsAndTokenIds: contractsAndTokenIds, enjinTokens: enjinTokens)
```

Comment:
Does the strong `self` reference here only cause a problem when the upstream code (`detectContractsAndTokenIds()` etc) don't emit any results/error?

Reply:
it causes an issue when its in progress, we ran function and when in same time switch servers, object should receive cancel as its parent dellocated